Fix: importer fails to import a basic obj in Blender 3.3.1 LTS

galvedro opened the pull request:
Trying to import basic geometry is failing in Blender 3.3.1 LTS. The script casts an exception saying that MeshVertex's `normal` property is readonly.

Sure enough, the property has become readonly. Moreover, writing to it was not doing much before the change either, apparently. See this discussion: [Make vertex normal property readonly](https://developer.blender.org/D14696)

This PR removes the offending lines. I have tested to import a number of objects of varying complexity in Blender 2.80 and Blender 3.3.1 LTS, including objects with normals pointing inwards. I have not noticed any problem with the change.

The discussion I linked above also talks about `calc_normals()`. I have not touched it, but you may want to take a look.

ian-m-carr commented:
I believe the vertex_normals collection is now the correct (non readonly) place to modify the normals. try the following replacement:

            for i, vertex_normal in enumerate(me.vertex_normals):
                vertex_normal = normals[i]
                
To remain backward compatible you need to do a version check and call this in the new or the old version in 2.8 blender!

galvedro commented:
Yes, that snippet works and, as you say, it breaks backwards compatibility with 2.80, so it needs to be handled.

But I am confused about whether or not this needs to be written by hand (I just don't know enough about 3D). Blender's documentation says:

> [vertex_normals](https://docs.blender.org/api/current/bpy.types.Mesh.html#bpy.types.Mesh.vertex_normals) - The normal direction of each vertex, defined as the average of the surrounding face normals.

If each vertex' normal is a function of the normals of the surrounding faces, doesn't Blender calculate them by itself, as described in the pull request I liked above, the one that made `normals` readonly in the first place?

ian-m-carr commented:
You are correct about the documentation! hadn't spotted that one, Odd that it does not except.

Interested to know if your removal of the lines results in import of the normals from the OBJ or are they just recalculated later?

todirbg commented:
How are the normals imported by wavefront obj importer? Maybe there is a clue on how it should be done in xp2b.  It produces better results than xp2b.

ian-m-carr commented:
Wavefront and fbx appear to set the normals via the mesh loops (split normals)
[wavefront:](https://github.com/blender/blender-addons/blob/8457285056e376827ce3863b880b9f01dcec50f8/io_scene_obj/import_obj.py#L705) which goes on to use `me.normals_split_custom_set` and [fbx](https://github.com/blender/blender-addons/blob/8457285056e376827ce3863b880b9f01dcec50f8/io_scene_fbx/import_fbx.py#L1198) which is setting them directly onto the me.loops.

there also seems to be a `me.normals_split_custom_set_from_vertices()` which may do what the x-plane code is trying to achieve taking a vertex normal set and applying it as 'split normals' see [this discussion](https://blenderartists.org/t/editing-normals/1233341) and the documentation: [normals_split_custom_set_from_vertices](https://docs.blender.org/api/current/bpy.types.Mesh.html#bpy.types.Mesh.normals_split_custom_set_from_vertices)

But calling me.calc_normals immediately after would presumably undo this work?

galvedro commented:
I just pushed a change. I added `auto_smooth`, set the normals with `normals_split_custom_set_from_vertices()`, and removed `calc_normals()`, as described in the post you linked. It does the trick, and it works in both 2.80 and 3.3.1.

I can see that the shading with which an OBJ is exported is respected when importing it again.

ian-m-carr commented:
Certainly an improvement! I made the smoothing an importer option in my implementation. There are some areas where the imported mesh still needs a little cleaning up. This panel from the C90B ends up with multiple normals on a number of the vertices. and looses the sharp edges, but a manual tidy up is possible! But the smoothing works well on the switches and other naturally smooth items.
![image](https://user-images.githubusercontent.com/7975054/197401497-dc49870d-b1aa-4a57-84fd-96336e033456.png)
detail of the corner:
![image](https://user-images.githubusercontent.com/7975054/197400805-6335e3d7-21e2-4b2c-a464-649ef6f56cb8.png)

ian-m-carr commented:
My feeling is that the order should be:

 ```
me.update(calc_edges=True)
me.normals_split_custom_set_from_vertices(normals)
```
(As per the wavefront importer) 

rather than
 ```
me.normals_split_custom_set_from_vertices(normals)
me.update(calc_edges=True)
```

Not entirely sure thtat the update calc_edges doesn't modify the normals again?

galvedro commented:
I do not see different results by changing the order. I have flipped them though, to be on the safe side.

One thing about smoothing. If I am understanding this right, `auto_smooth` has to be enabled in the object for Blender to do something with the custom normals, otherwise it won't use them and will do flat shading instead.

ian-m-carr commented:
Agree, I made it an option as some objects like the one above import better if flat shaded, that panel for instance shows rounded edges and wrinkled surface if imported smooth, but looks correct if imported flat. But on the whole importing everything smooth and hand fixing them after looks to be the better route. The fact that we are making a choice for all objects in a file, tends to make smooth the better bet. 

What we could do with is a **simple** automated way to recognise and mark the sharp edges from the imported normals.

ian-m-carr commented:
Interestingly the export code in xplane_mesh.py looks at the smooth flag (on the face) and either exports using the split normals or the face normals.

` for tmp_face in tmp_faces:
                    # To reverse the winding order for X-Plane from CCW to CW,
                    # we iterate backwards through the mesh data structures
                    for i in reversed(range(0, 3)):
                        index = tmp_face.indices[i]
                        vertex = xplane_helpers.vec_b_to_x(mesh.vertices[index].co)
                        normal = xplane_helpers.vec_b_to_x(
                            tmp_face.split_normals[i]
                            if tmp_face.original_face.use_smooth
                            else tmp_face.normal
                        )`

ian-m-carr commented:
which normals are you changing, vertex or face? if you export non-smooth it will ignore changes to the vertex normals and use the face ones.

As far as I can see the split normals are correct in all cases, can't quite see why it chooses the alternatives.

flat shading the split normals are following the face normals
![image](https://user-images.githubusercontent.com/7975054/197585563-ffe41571-18ce-489f-ba14-1387a2066101.png)
smooth shading (object) the split normals average the two faces
![image](https://user-images.githubusercontent.com/7975054/197586069-fe156432-7453-4265-b57c-cff77aa55a16.png)

Auto-smooth is the same as flat in the default case
![image](https://user-images.githubusercontent.com/7975054/197586289-dc8cea80-25a7-407a-9baf-7c3fc69c6a1a.png)

galvedro commented:
I agree. Changing the exporter to always use split normals gets this test scene exported correctly, and the importer also gets it right back into Blender. The scene has three objects, one with smooth shading, one with flat shading, and one with auto_shading and a Weighted Normal modifier applied.

This is the imported result:
<img width="1261" alt="normals_test" src="https://user-images.githubusercontent.com/11858803/197586175-b4e30889-6ad5-4b24-bd0e-218f8321e27e.png">

It does look like the exporter could just use the split normals. Could this have been an issue in an older version of Blender? I am testing with 3.3.1 at the moment...

ian-m-carr commented:
I spotted the behaviour in the exporter earlier and thought it was odd to choose not to use the split normals always. I have been using 2.8.3 for my testing as I had hit the animation keyframe issue you had reported (Thanks for the fix by the way!!!)

It would be worth running a python test over the meshes you have there and look at the face.smooth flag and the object smooth flag, see how they are set in the auto-smooth case.

ian-m-carr commented:
So first two would use face normals and the only the third will use split normals in the original code.

I think the "always use split normals" sounds like the right way to go here?

ian-m-carr commented:
I can try it out in 2.8.3, and 2.9 tomorrow see what the results are there

galvedro commented:
Exactly, that's why the more advanced vertex normal edits are not exporting correctly.

Not sure how you want to handle the changes. If you want I can push the change to the exporter in this branch and maybe rescope the PR to something like "Fix: vertex normal handling" or something like that.

ian-m-carr commented:
That's up to the devs, I am just an interested party like yourself, trying to get a working toolset! But the rescope sounds like it would make sense to me!

galvedro commented:
😀 Now I understand why you're playing with your fork instead of merging my PRs! Lol!

ian-m-carr commented:
Yeah, with the latest tweaks here it's looking pretty good. From the default import,
 normals are looking much cleaner!:
![image](https://user-images.githubusercontent.com/7975054/197721857-96086035-8e67-4011-be9e-223549cca76d.png)

The fuel panel is coming in with sharp edges and without surface defects, without manual modification!
![image](https://user-images.githubusercontent.com/7975054/197723181-3cefd644-d247-43bd-a990-ecba5923e4a3.png)

I think it's now correctly representing the normals from the imported file!

and thats in 3.3.1!

I Just need to check the export isn't now bloating the files again. my fix to apply smoothing to all the faces reduced the large number of similar normals at the vertices.

ian-m-carr commented:
Ok export file size is better than the 4.2.0-alpha original code but not as small as my 'all surfaces smooth' 

```
| 1.46MB | POINT_COUNTS	12936	0	0	111594 | original file
| 1.23MB | POINT_COUNTS	10618	0	0	111108 | my export (all faces marked smooth in test_creation_helpers create_object)
| 3.13MB | POINT_COUNTS	42384	0	0	111108 | export from this discussion (mesh marked auto-smooth)
| 9.97MB | POINT_COUNTS	111594	0	0	111594 | original importer from io_xplane2blender_4_2_0-alpha
```

So significantly better than the old code (3*), but produces 42384 unique vertex/normal pairs rather than the 10618 that my smooth all surfaces ends up with. (But my version requires manual intervention on certain sharp areas like that flat panel)

Overall it's a trade off on each file, I think for now I am going to implement both options in my version adding them as separate import options.

**Note** my version also has the global optimisation fix I proposed so your figures may differ!

ian-m-carr commented:
understand why mine is reducing the count, yes the smoothing reduces the number of split vertices.

It's why the solution here increases the number of normals by a factor of 4 over the original file, when it should now be applying the set of normals we just imported! My only thought is that the validate called in the importer is resulting in the additional normals. 

I am still seeing many discrete normals at the coincidence points of suposedly flat faces for instance this one:

![image](https://user-images.githubusercontent.com/7975054/198061675-f51a9921-ba0f-4eb3-8f84-2d894a259ec0.png)

My approach certainly cleans these up.

Still it's better than the original code which increases the number of normals by a factor of nearly 10!

galvedro commented:
> when it should now be applying the set of split-normals we just imported

The call to
```py
me.normals_split_custom_set_from_vertices(normals)
```
is splitting the single normal per vertex that is coming from the OBJ into multiple normals per vertex. It could be that if the faces sharing that vertex are not dead on coplanar, the split normals may deviate from the original normal in the file, giving us a bouquet of normals like the one in your screenshot.

ian-m-carr commented:
It's something like that, I think the only way to tell is to trawl the input file, but with 111000 points it's going to take me a while to isolate the right ones :-)

I get the feeling that a smarter export optimisation that considers similarity (close angular match of the split normal vectors) rather than the current "exact match" of the normals would resolve the issue at the expense of a little speed on the export.

todirbg commented:
Great discussion here! I don't know a thing about python and blender plugins and I am not sure I understand what you are talking about, but a while ago I made an obj8 importer for armor paint in haxe and I got some understanding on the obj8 format itself. 
The first thing my code does while importing is deoptimize obj8, if it was ever optimized! Any optimization on the mesh goes away and then the import can proceed. I import the normals as they are just scaling and transforming the way armor paint mesh data expects them. The mesh imports perfectly! Armor paint can then export to wavefront obj and then if I improt that wavefront in blender it is almost perfect. The wavefront importer creates custom normals data with sharp edges and smooth shading, but it looks almost perfect. So I wander why not you just parse the obj8 into vertex data and than use the wavefront importer functions to make blender mesh out of it?
Here is the base of my importer if it can help you in any way https://github.com/todirbg/armorbase/blob/a44bed8813c90114b56a55612c5982d88768e5f4/Sources/arm/format/Obj8Parser.hx
It splits the mesh into objects including some of the animation data and then does some transformations with it to make each object appear in its "rest pose".

ian-m-carr commented:
Cheers, I'll take a look see what we can learn!

galvedro commented:
> I get the feeling that a smarter export optimisation that considers similarity (close angular match of the split normal vectors) rather than the current "exact match" of the normals would resolve the issue at the expense of a little speed on the export.

It smells like that, yep.

bsupnik commented:
Hey Gang,

First, thank you for the PR and for taking a look at this.

I think the main thing that I need to talk with the internal art team about is whether truly ad-hoc normals are something we should support.

The case in favor of them would be perfect round trip IO with the importer/exporter, but that's kind of a weird goal - we view the importer as a lifeboat to get old content authored in other 3-d tools that the community has heavily relied upon (Blender 2.49, AC3D) into modern Blender with animations intact to protect investment in modeling.

But in this case, the intent of the original normals might be lost, and a "split all the time so we can be perfectly like the old file" eats a pretty big cost (a multiplier on totla vertex count and a loss of reuse, as well as smooth-shading artifacts).

It looked to me like the intent in Blender 3.x is to have normals be a derived, calculated quantity that Blender computes and we consume, hence a MeshVertex's normal is no longer writable. If this is the case, are we petting the cat backward by managing our noramls by hand?

cheers
Ben
